### ISSUE TITLE
Rearrange / drag tabs/panes into windows and vice-versa

### ISSUE THREAD

**Issue by nipunn1313** [warpdotdev/warp#466]
### Describe the bug

Similar to #58 - would love to use drag-and-drop to convert tabs into windows and vice versa.

Open to other UI mechanism ideas - but drag-and-drop is the one I'm used to for this functionality. Something in cmd-p or in the tab menu could work too.

### To Reproduce

Steps to reproduce:
2. Click on '...'
3. Scroll down to '...'
4. See error '...'

**Comment by elviskahoro** [comment]
Linking:
- #58

**Comment by elviskahoro** [comment]
Linking:
- #398

**Comment by dannyneira** [comment]
Thanks for this feature request! Please :+1:  and subscribe to this request as it helps us gauge interest. We will post any updates here.

**Comment by bsramin** [comment]
same of https://github.com/warpdotdev/Warp/issues/623

**Comment by thatguyintech** [comment]
I'd love to have this feature! I change around my monitor resolution and orientation occasionally so being able to change my warp tab vs window layout helps a bunch for quickly transitioning between layouts.

**Comment by khanakia** [comment]
Is there any update on this this is only the reason stopping to use Warp on my day-to-day activities.

**Comment by sakers** [comment]
I'm also back to using iterm2 as I really need this functionality.

**Comment by aidan-gibson** [comment]
Absolutely batshit this doesn't exist. I love Warp but come on.

**Comment by theDanielJLewis** [comment]
I'd love to have this, too! I'd especially love the ability to drag any of these tabs into a split, and drag something out of a split into a tab or separate window. (Yes, iTerm can do both of these things.)

**Comment by RorzGonzalez** [comment]
+1 ... not thrilled that I'm forced to a new window if I choose to use launch configs. I prefer a quick way to access different servers in a new tab within one window. Currently using the warp workflows which works fine but I prefer the launch_configs for server connections.

**Comment by theDanielJLewis** [comment]
Yes, @RorzGonzalez. It _would_ be nice if configs had the option to launch as new tabs in an existing window.

**Comment by micronax** [comment]
This is really a crucial feature and tbh keeps me from using Warp on a day-to-day base. :(

**Comment by sakers** [comment]
[like]  Sean Akers reacted to your message:
________________________________

**Comment by vehvis** [comment]
Sad to see four years have passed and this no-brainer feature is still not done.

**Comment by claytonolley** [comment]
came here to say the same... guess I'll put this app back on the shelf for a while and check back in another 6 months again...

**Comment by sakers** [comment]
[like]  Sean Akers reacted to your message:
________________________________

**Comment by sakers** [comment]
[like]  Sean Akers reacted to your message:
________________________________

**Comment by bedge** [comment]
Use case:
1. Split a window to get a shell to test the thing one is editing in the other pane.
2. Realize that one needs more screen real estate for both panes, but needs to see then concurrently on different monitors.
3. Would like to take a split pane and move to a new window.
4. ... and back, so when the "need to see both, in full screen on different monitors" ends, one can re-attach to the previous pane's split.

**Comment by theDanielJLewis** [comment]
@twardoch, part of that is possible, but weirdly so.

If you have two tabs open and want to merge them into a single tab, you can do that but only if the tab you're pulling from has more than one pane. Then you can grab that pane and drop it into the other tab. But, at least that I can tell, you can't do this if the other tab contains only one pane.

**Comment by bedge** [comment]
Thanks @twardoch, good to know.

What I find quite odd is that it took me a long time to even figure out how to open a second window - by running `warp` a second time, and even then, they are totally distinct entities.
I realize I'm focusing on an area that is not warp's main focus, but I'm looking at whether I can make this a daily driver first. If not, there's little sense in burning time on the full learning curve.

What I'm missing already from iterm2 & windows terminal is the ability to pop tabs out into new windows, then rejoin.
That would be less of an issue if there was some form of tmux/screen server to share virtual sessions. Using tmux here would really be defeating the purpose if warp.
Granted I haven't played with "teams" and sharing yet. 
Maybe that's next, as perhaps that's the solution. I often feel like a one man team :)

**Comment by theDanielJLewis** [comment]
@bedge:

> What I find quite odd is that it took me a long time to even figure out how to open a second window - by running warp a second time, and even then, they are totally distinct entities.

Cmd-N on macOS, and I assume Ctrl-N on Windows does the same thing.

**Comment by bedge** [comment]
> [@bedge](https://github.com/bedge):
> 
> > What I find quite odd is that it took me a long time to even figure out how to open a second window - by running warp a second time, and even then, they are totally distinct entities.
> 
> Cmd-N on macOS, and I assume Ctrl-N on Windows does the same thing.

That was my first thing to try :)

Ctrl-N does nothing.

There's not even a "new window" in the "keyboard shortcuts" settings:

![Image](https://github.com/user-attachments/assets/6e387070-6152-48e9-91c3-b5953040e33f)

What's that shortcut label on Mac? Maybe I'm looking for the wrong name?

**Comment by theDanielJLewis** [comment]
It's right in the File menu.

<img width="247" alt="Image" src="https://github.com/user-attachments/assets/963ea882-8a7f-4de6-ae80-ada1396a823e" />

But it's apparently not a changeable shortcut because there's no "new window" listing under the keyboard shortcuts settings.

**Comment by twardoch** [comment]
> What I'm missing already from iterm2 & windows terminal is the ability to pop tabs out into new windows, then rejoin.

Yeah. That's really badly needed!

**Comment by sviatoslav-lebediev** [comment]
In case somebody needs to change New Window mapping, I did it on the macOS level

![Image](https://github.com/user-attachments/assets/58f210f8-0e13-486b-95b3-2c092443859d)

**Comment by pieterza** [comment]
Literally can't use it due to this missing feature.  My workflow has been trained to drag tabs between monitors, while Warp is nice, no feature sells me on having to retrain decades of muscle memory.

Assuming it's an issue with your GUI library choice, which is why this is still open?

**Comment by Repulser** [comment]
Super needed! my top feature request for warp.

**Comment by rgaufman** [comment]
Indeed, I really miss this coming from iterm2, I often drag a tab from 1 window to another.  As part of that, also dragging to create a pane, right now I have to split the pane in both tabs, then drag, then close the panes I never needed in the first place, but had to create to see the pane drag handle.

**Comment by sakers** [comment]
They keep adding lots of new buzzword features and AI fluff but don't add this basic functionality ? I don't understand why this pretty basic feature is so hard for them.

**Comment by xmarkclx** [comment]
Maybe they think it won't generate any money? @sakers 

But atm. the lack of having this feature is chasing away potential users, and is blocking their dominance as a good replacement terminal app vs the default ones from each OS.

**Comment by sakers** [comment]
Possibly @xmarkclx. However, the lack of this feature has sent me back to iTerm2 and ghostty. There's no way I'd pay for Warp until it has this feature. Not even using it for free currently.

**Comment by Nimble-Jack** [comment]
@dannyneira any update on this? it's pushing 4 years being open, regarding the 'pop-out' of tabs

**Comment by dannyneira** [comment]
No updates at this time, and it's not on the roadmap as far as I'm aware. Please continue to 👍 the [top request](https://github.com/warpdotdev/Warp/issues/466#issue-1066641732) so we can track interest in this feature.

**Comment by sakers** [comment]
I've given up on Warp. It's become a P.I.T.A to use. Too intrusive AI. The lack of tear-off tabs in ridiculous in this day-and-age. iTerm and Ghostty are doing the job. I don't care about AI features. Warp has lost its way.

**Comment by elliottash** [comment]
+1 on this request, features like this will distinguish warp from the 10 other CLI agents.

**Comment by dicebattle** [comment]
I think warp is awesome in all other features. but because of ONLY THIS FEATURE, I installed ghostty today.
I'll try to use both of them.

**Comment by joseph-mcallister** [comment]
+1

**Comment by Mustafa-Esoofally** [comment]
How is this still not added?

**Comment by pieterza** [comment]
> How is this still not added?

I'd understand if this was shitty/lazy electron that perhaps makes it difficult to do pure desktop.

But this is Rust.  Perhaps the functionality doesn't exist yet in the core lib or something.

Or the Warp devs just don't give a shit.

**Comment by elliottash** [comment]
i think this works now -- dragging panes to tabs and vice versa.

**Comment by prodrammer** [comment]
### Editor groups in VS Code. Created using drag/drop.

<img width="966" height="1043" alt="Image" src="https://github.com/user-attachments/assets/62f07f8d-0140-40c9-af60-89d30bc70471" />

**Comment by gabrielbacha** [comment]
Any update on this? It's been a month since the last update @captainsafia @leesa-ui-w

**Comment by theDanielJLewis** [comment]
> Nah, they keep it. Unsubscribing from this issue and moving on with life. Epic fail.

Me too. I'm moving to Cmux.

Its drag-and-drop isn't fully there, but you _can_ drag-and-drop from one window to another. And you can right-click a tab to move it from one workspace to another.

**Comment by captainsafia** [comment]
@gabrielbacha Thanks for pinging on this thread! There's no updates at the moment, other than this is still in-progress. I know this is not the ideal response, but this feature is fairly involved and we want to make sure that the implementation is polish across all the different tab/window/pane permutations. 

We're using this issue as the canonical tracking issue and will be sure to share updates as they come.

**Comment by pelmered** [comment]
> https://x.com/BHolmesDev/status/2032587085689856367
> 
> big if true

Too little too late for me. I really enjoyed using Warp, but there is no excuse for taking this long for a feature like this. Now it has been in review for a month as well. 
A few weeks ago I moved to Cmux and this was the main annoyance. I don't think this is not enough for me to come back now that I have switched and learned a new terminal.

**Comment by RonAmihai** [comment]
Please implement that feature already; it is extremely basic.
I have used Warp consistently until that missing feature made me switch to ghostty (cmux also lacks those features)

Once you add that feature (including "merge all windows", splitting tabs from windows into new windows, reordering tabs, etc.), I am considering moving back. It's the main missing piece in that terminal.

**Comment by GeoJunkie** [comment]
Looks like there's a PR for this: https://github.com/warpdotdev/warp/pull/9275

**Comment by BudhirajaMadhav** [comment]
For folks subscribed to this thread: the tab↔window slice of this is in flight at #9275 — Chrome-style cross-window tab drag (detach to new window, attach to another window's tab bar, multi-cycle reattach), behind the `DragTabsToWindows` feature flag, opened 2026-04-28, CI green.

The complementary "merge a tab into another tab as a split pane" capability — the inverse of today's pane→tab tear-out, which several comments here have asked for — isn't in #9275's scope and is tracked separately at #8959.

**Comment by jeremyronking** [comment]
Can we still not move tabs between windows? I was searching for an existing issue and found this.  This is common workflow for Mac users. Please prioritize for us!

**Comment by GeoJunkie** [comment]
Apparently this is fixed now! I just randomly dragged a tab and it allowed me to drag it between windows 🥳

**Comment by hellosagar** [comment]
@GeoJunkie but now we moved to Ghostty now lol, but thanks for sharing :) will do try it out

**Comment by trliner** [comment]
Confirmed it's working for me too. 🚀

**Comment by jeremyronking** [comment]
Works for tabs, but not tab groups 😢

**Comment by acarl005** [comment]
Closing this out as the core ask is delivered. Still some other TODOs but those have their own issues

•  Tab to pane-split merging: #8959
•  Tab groups across windows: #14152
